File: 13-10-2025_Maximum Non-Adjacent Nodes Sum.py
def solve(root):
    if not root:
        return (0, 0)  # (include, exclude)

    left = solve(root.left)
    right = solve(root.right)

    include = root.data + left[1] + right[1]
    exclude = max(left) + max(right)

    return (include, exclude)

def getMaxSum(root):
    include, exclude = solve(root)
    return max(include, exclude)

File: 13-10-2025_Maximum Non-Adjacent Nodes Sum.cpp
pair<int, int> solve(Node* root) {
    if (!root) return {0, 0};  // {include, exclude}

    auto left = solve(root->left);
    auto right = solve(root->right);

    int include = root->data + left.second + right.second;
    int exclude = max(left.first, left.second) + max(right.first, right.second);

    return {include, exclude};
}

int getMaxSum(Node* root) {
    auto ans = solve(root);
    return max(ans.first, ans.second);
}
